wtxmgr/tx_test.go

```diff
@@ -1364,3 +1364,217 @@ func TestRemoveUnminedTx(t *testing.T) {
+// mempool should be removed from the wallet's store.
+func TestInsertMempoolDoubleSpendConfirmedFirstTx(t *testing.T) {
+	t.Parallel()
+	testInsertMempoolDoubleSpendTx(t, true)
+}
+
+// TestInsertMempoolDoubleSpendConfirmedFirstTx ensures that when a double spend
+// occurs and both spending transactions are present in the mempool, if the
+// second spend seen is confirmed, then the first spend transaction within the
+// mempool should be removed from the wallet's store.
+func TestInsertMempoolDoubleSpendConfirmSecondTx(t *testing.T) {
+	t.Parallel()
+	testInsertMempoolDoubleSpendTx(t, false)
+}
+
+// TestInsertConfirmedDoubleSpendTx tests that when one or more double spends
+// occur and a spending transaction confirms that was not known to the wallet,
+// then the unconfirmed double spends within the mempool should be removed from
+// the wallet's store.
+func TestInsertConfirmedDoubleSpendTx(t *testing.T) {
+	t.Parallel()
+
+	store, db, teardown, err := testStore()
+	defer teardown()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dbtx, err := db.BeginReadWriteTx()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer dbtx.Commit()
+
+	ns := dbtx.ReadWriteBucket(namespaceKey)
+
+	// We'll start off the test by creating a new coinbase output at height
+	// 100 and inserting it into the store.
+	b100 := BlockMeta{
+		Block: Block{Height: 100},
+		Time:  time.Now(),
+	}
+
+	cb := newCoinBase(1e8)
+	cbRec, err := NewTxRecordFromMsgTx(cb, b100.Time)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := store.InsertTx(ns, cbRec, &b100); err != nil {
+		t.Fatal(err)
+	}
+
+	// Then, we'll create three spends spends from the same coinbase output.
+	// The first two will remain unconfirmed, while the last should confirm
+	// and remove the remaning unconfirmed from the wallet's store.
+	firstSpend := spendOutput(&cbRec.Hash, 0, 5e7, 5e7)
+	firstSpendRec, err := NewTxRecordFromMsgTx(firstSpend, time.Now())
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := store.InsertTx(ns, firstSpendRec, nil); err != nil {
+		t.Fatal(err)
+	}
+	secondSpend := spendOutput(&cbRec.Hash, 0, 4e7, 6e7)
+	secondSpendRec, err := NewTxRecordFromMsgTx(secondSpend, time.Now())
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := store.InsertTx(ns, secondSpendRec, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	// At this point, we should see both unconfirmed transactions within the
+	// store.
+	unminedTxs, err := store.UnminedTxs(ns)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(unminedTxs) != 2 {
+		t.Fatalf("expected 2 unmined txs, got %d", len(unminedTxs))
+	}
+
+	// Then, we'll insert the confirmed spend at a height deep enough that
+	// allows us to successfully spend the coinbase output.
+	coinbaseMaturity := int32(chaincfg.TestNet3Params.CoinbaseMaturity)
+	bMaturity := BlockMeta{
+		Block: Block{Height: b100.Height + coinbaseMaturity},
+		Time:  time.Now(),
+	}
+	confirmedSpend := spendOutput(&cbRec.Hash, 0, 3e7, 7e7)
```

Comment:
you can add a third unconfirmed spend to the store that spends a second coinbase output. Then you make this `confirmedSpend` spend both coinbase outputs in the same tx.

Reply:
Fixed.